Update noda time plugin to support bcl types by forwarding these to the existing bcl type handlers.

davidroth opened the pull request:
This is a fix so that noda types can co-exist with bcl types until clean support for multiple type handlers will be implemented.
See issue #2439 for details.

YohDeadfall commented on src/Npgsql.NodaTime/DateHandler.cs:
```diff
@@ -65,5 +71,29 @@ public override void Write(LocalDate value, NpgsqlWriteBuffer buf, NpgsqlParamet
             buf.WriteInt32(totalDaysSinceEra - 730119);
         }
+
+        protected internal override int ValidateObjectAndGetLength(object value, ref NpgsqlLengthCache lengthCache, NpgsqlParameter parameter)
```
There are too many methods in type handlers which makes it hard to understand where is the actual entry point, so I perfectly understand your confusion here (:

What you should touch is internal methods having the `TAny` generic parameter: `Read<TAny>`, `ValidateAndGetLength<TAny>` and `WriteWithLengthInternal<TAny>`. These methods are responsible for strongly typed read and write operations and they are non-allocating. Internally they call corresponding methods of `INpgsqlTypeHandler<T>` if the class implements `INpgsqlTypeHandler<TAny>`.

Other three entry points are non-generic methods `ReadAsObject`, `ValidateObjectAndGetLength` and `WriteObjectWithLength`. Since they work with objects they allocate memory when the parameter or the default type is a value type because of boxing as you surely understand. Also the non-generic write is slower that the generic version because it calls a delegate pointing to `WriteWithLengthInternal<TAny>` in a generated `if (value is T) WriteWithLengthInternal<T>(/* parameters */);`.

Therefore, to support all scenarios all generic and non-generic methods should check the argument and decide what handler should be used, but this costs us an unmaintainable code (it really smells, take a look at #3092). Plus this interception decreases performance for NodaTime types which we don't want. So let's implement interfaces that the corresponding type handlers for BCL have and delegate handling in method implementations like you did in this handler. The entry point methods should be removed.

roji replied:
I think this is simpler... The entry points that work with objects, as well as those that are generic, are designed to be implemented by type handlers which handle an open set of types in some way. For example, the JSON type handler can handle any .NET type, and the array type handler handles arbitrarily dimensioned arrays (`int[]`. `int[,]`, `int[,,]`...).

In the simpler case, type handlers with a closed set of supported types - like this one - should just implement INpgsqlTypeHandler (or INpgsqlSimpleTypeHandler) for each supported type; no need for anything more complicated. So just like you've implemented `INpgsqlSimpleTypeHandler<NpgsqlDate>`, the type handler can implement `INpgsqlSimpleTypeHandler<DateTime>`, and the methods coming from that should make everything work easily.

YohDeadfall replied:
You simplified what I tried to say, it looks much clear (:

roji replied:
Oh great :)

I agree the type handler system right now is a bit... complicated (maybe even convoluted). It evolved this way over several years. I actually think there's an ok logic - the typical/regular type handler just implements the interface for the types it wants to support, and extends NpgsqlTypeHandler for the "default" type. But it's true there's too much indirection going on, making it both confusing and less performant.

YohDeadfall commented on src/Npgsql.NodaTime/NpgsqlNodaTimeExtensions.cs:
```diff
@@ -61,7 +62,7 @@ public static INpgsqlTypeMapper UseNodaTime(this INpgsqlTypeMapper mapper)
                     PgTypeName = "interval",
                     NpgsqlDbType = NpgsqlDbType.Interval,
-                    ClrTypes = new[] { typeof(Period) },
+                    ClrTypes = new[] { typeof(Period), typeof(TimeSpan) },
```
Forgot about `NpgsqlTimeSpan`.

roji replied:
As a general rule, it's fine to leave out all the Npgsql date/time types (e.g. NpgsqlDate) - we're considering obsoleting/removing them anyway. Or you can delegate them as well, up to you (I doubt anyone actually needs them, especially when using the Nodatime plugin).

YohDeadfall commented on src/Npgsql.NodaTime/TimeTzHandler.cs:
```diff
@@ -1,8 +1,10 @@
 using Npgsql.BackendMessages;
 using Npgsql.PostgresTypes;
 using Npgsql.TypeHandling;
+using BclDateTimeOffsetHandler = Npgsql.TypeHandlers.DateTimeHandlers.TimeTzHandler;
```
Let's rename this to `BclTimeTzHandler` like in other cases.

YohDeadfall commented on src/Npgsql.NodaTime/TimestampTzHandler.cs:
```diff
@@ -1,9 +1,11 @@
 using Npgsql.BackendMessages;
 using Npgsql.PostgresTypes;
 using Npgsql.TypeHandling;
+using BclDateTimeOffsetHandler = Npgsql.TypeHandlers.DateTimeHandlers.TimestampTzHandler;
```
And here too:
```suggestion
using BclTimestampTzHandler = Npgsql.TypeHandlers.DateTimeHandlers.TimestampTzHandler;
```

YohDeadfall reviewed (changes requested):
Thank you for submitting the fix! Let's do review now...

roji commented on test/Npgsql.PluginTests/NodaTimeTests.cs:
```diff
@@ -364,5 +363,55 @@ protected override NpgsqlConnection OpenConnection(string? connectionString = nu
         #endregion Support
+
+        [Test, Description("Makes sure noda time type handlers are forwarded to BCL type handlers.")]
+        public void BclTypeHandlerForwarding()
```
I think this test should be redundant - the specific NodaTime tests should do all the testing. For example, test Interval above already checks that values can be read as a BCL TimeSpan - it just needs to add another parameter to test writing of TimeSpan as well.

YohDeadfall replied:
Since it's a workaround to support both type handlers at once let's add the provider specific types.

davidroth replied:
You mean I should remove this additional test method and update the specific test methods accordingly?

roji replied:
Yeah, that's my view.

There's also the question of if you want to support the provider-specific types (NpgsqlDate) in addition to the BCL types (DateTime). I definitely don't think this is mandatory, but if you want to you can again add testing in the existing tests.

roji reviewed (commented):
Looks great, but see note about simplifying how the type handlers work.

davidroth commented:
Thanks guys for the feedback! I will push an update probably tomorrow.

YohDeadfall commented on src/Npgsql.NodaTime/TimeTzHandler.cs:
```diff
@@ -17,9 +18,13 @@ public override NpgsqlTypeHandler<OffsetTime> Create(PostgresType postgresType,
     }
 
-    class TimeTzHandler : NpgsqlSimpleTypeHandler<OffsetTime>
+    sealed class TimeTzHandler : NpgsqlSimpleTypeHandler<OffsetTime>, INpgsqlSimpleTypeHandler<DateTimeOffset>,
+                                    INpgsqlSimpleTypeHandler<DateTime>, INpgsqlSimpleTypeHandler<TimeSpan>
```
nit: Let's align them.

YohDeadfall commented on src/Npgsql.NodaTime/TimestampTzHandler.cs:
```diff
@@ -22,10 +23,11 @@ public override NpgsqlTypeHandler<Instant> Create(PostgresType postgresType, Npg
 
-    class TimestampTzHandler : NpgsqlSimpleTypeHandler<Instant>, INpgsqlSimpleTypeHandler<ZonedDateTime>,
-                               INpgsqlSimpleTypeHandler<OffsetDateTime>
+    sealed class TimestampTzHandler : NpgsqlSimpleTypeHandler<Instant>, INpgsqlSimpleTypeHandler<ZonedDateTime>,
+                               INpgsqlSimpleTypeHandler<OffsetDateTime>, INpgsqlSimpleTypeHandler<DateTimeOffset>
```
nit: And here let's align interfaces too.

YohDeadfall reviewed (changes requested):
Much better, thanks! What you forgot to do is to register CLR types for `time` types in the `UseNodaTime` method.

davidroth commented:
> Much better, thanks! What you forgot to do is to register CLR types for `time` types in the `UseNodaTime` method.

I'll continue tomorrow. Also need to finish work on unit test.

davidroth commented:
> Much better, thanks! What you forgot to do is to register CLR types for `time` types in the `UseNodaTime` method.

Since a CLR Type can only be used in a single mapping, I am not sure wich CLR types you mean? 
(`TimeSpan` is already used for `NpgsqlDbType.Interval`)

davidroth commented:
Ok guys, I think I am done for now and this should be ready for final review.

- Removed non generic overrides and implement `INpgsqlSimpleTypeHandler`s instead.
- Updated `NodaTimeTests` so that existing test methods check against bcl types
- nip cleanup

Regarding the "InternalsVisibleTo" question:

This is needed because BCL type handler Ctors/Methods are internal. Since i need to delegate to them, InternalsVisibleTo is needed.

Hope its ok now. Please let me know if something is missing.

davidroth commented on src/Npgsql.NodaTime/TimestampHandler.cs:
```diff
@@ -167,5 +173,14 @@ internal static void WriteDouble(Instant instant, NpgsqlWriteBuffer buf)
+
+        void INpgsqlSimpleTypeHandler<DateTime>.Write(DateTime value, NpgsqlWriteBuffer buf, NpgsqlParameter parameter)
+            => ((INpgsqlSimpleTypeHandler<DateTime>)_bclHandler).Write(value, buf, parameter);
     }
```
You might be wondering why I have an explicit cast here.  This is because there is an implicit conversion from `DateTime` to `NpgsqlDateTime`. Therforce c#s overload resolution always picks `NpgsqlDateTime` methods in the BCL handler. Since there is special handling for DateTime.MaxValue/MinValue (convertInfinityDateTime), this matters.

roji replied:
Thanks for the explanation, another problematic aspect of the provider-specific types... We should consider removing the implicit cast operators (unless we remove the types entirely...)

davidroth commented:
One more thing: I resolved all previous discussions here to cleanup for the next round. Hope thats ok ;-)

davidroth commented:
Question: Once this is merged, it probably makes sense to update noda time unit tests in "efcore.pg" accordingly? E.x. I would add CLR datetime types here: https://github.com/npgsql/efcore.pg/blob/dev/test/EFCore.PG.NodaTime.FunctionalTests/NodaTimeQueryNpgsqlTest.cs#L431

davidroth commented on src/Npgsql/Properties/AssemblyInfo.cs:
```diff
@@ -22,6 +22,13 @@
 "7aa16153bcea2ae9a471145624826f60d7c8e71cd025b554a0177bd935a78096" +
 "29f0a7afc778ebb4ad033e1bf512c1a9c6ceea26b077bc46cac93800435e77ee")]
 
+[assembly: InternalsVisibleTo("Npgsql.NodaTime, PublicKey=" +
```
This is needed because BCL type handler Ctors/Methods are internal. Since i need to delegate to them, InternalsVisibleTo is needed.

> I don't see any references to internals in NodaTime tests.

Its not for the `Npgsql.PluginTests` assembly, its for `Npgsql.NodaTime´ assembly.

roji replied:
I think it's best to make these ctors/methods public rather than introducing IVT.

The type handlers are currently somewhat inconsistent with public/internal - we've made public certain things because we needed them, but not others. As a general rule, the type handers should be a sort of "public internal" (similar to how EF Core manages things): they should be public for easy extension, but without any guarantees against breaking changes (at least in the near future).

roji reviewed (approved):
Thanks, this looks great to me! See only the note about InternalsVisibleTo which I'd rather we avoided.

> Once this is merged, it probably makes sense to update noda time unit tests in "efcore.pg" accordingly? E.x. I would add CLR datetime types here: https://github.com/npgsql/efcore.pg/blob/dev/test/EFCore.PG.NodaTime.FunctionalTests/NodaTimeQueryNpgsqlTest.cs#L431

Yeah, that's a good idea - but we'd have to make the EFCore.PG depend on a newer version of Npgsql for that to work. At some point I have to start doing this anyway - the upcoming EFCore.PG 5.0 should take a dependency on Npgsql 5.0, which we should release at the same time. So it would be good if you could open an issue for these tests.

davidroth commented:
Ok, I have removed IVT and made the needed ctors and Read method public.

YohDeadfall reviewed (approved):
Thank you!

YohDeadfall commented:
@roji, should it be backported to 4.1.5? It breaks nothing, but relaxes restrictions and could make users happier.

roji commented:
Thanks @davidroth! I'll look into getting the EF Core provider running on this so we can add tests etc.

@YohDeadfall good point. We should usually try to do only bugfixes in patches, but I agree this is very low risk and seems to be causing issues to quite a few people. Backported via 64827facb823fe1eae4d088fc3d132c6889bca85.

willdean commented:
Thanks very much to you all for working on this - we have been blocked from using NodaTime with PG/EF until now, and every time we have a DateTime-related UTC vs Local screw-up I've longed for this to be sorted.  Putting back it into 4.x is the cherry on top of the icing on top of the cake.
